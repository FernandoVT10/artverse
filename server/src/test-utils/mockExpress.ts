import { NextFunction } from "express";

type Parameters = {
  bodyData?: { [key: string]: any };
};

type ExpressMocked = {
  req: any;
  res: any;
  next: NextFunction;
};

export default function (params?: Parameters): ExpressMocked {
  const req = {
    body: params?.bodyData || {},
  };

  const res: any = {};

  res.json = jest.fn();
  res.status = jest.fn(() => res);

  const next = jest.fn();

  return {
    req,
    res,
    next,
  };
}
