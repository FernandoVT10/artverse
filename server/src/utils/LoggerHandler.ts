/* eslint @typescript-eslint/no-explicit-any: "off" -- In this file we're gonna need to use the 'any' type */
import logger from "@config/logger";

class LoggerHandler {
  static logError(error: any) {
    logger.error(error);
  }

  static logWarn(warn: any) {
    logger.warn(warn);
  }

  static logInfo(infoMessage: any) {
    logger.info(infoMessage);
  }
}

export default LoggerHandler;
