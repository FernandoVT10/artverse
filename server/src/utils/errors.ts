export class ServerError extends Error {
  public statusCode: number;

  constructor(message?: string, statusCode = 500) {
    super(message);
    this.statusCode = statusCode;
  }
}

export class ValidationError extends Error {
  public statusCode: number;
  public field: string;

  constructor(message: string, field: string) {
    super(message);
    this.statusCode = 400;
    this.field = field;
  }
}
