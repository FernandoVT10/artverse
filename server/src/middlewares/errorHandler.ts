import type { ErrorRequestHandler } from "express";

import { ServerError, ValidationError } from "@utils/errors";

const DEFAULT_ERROR_MESSAGE =
  "There was an error trying to process your request. Please try again later.";

export default function errorHandler(): ErrorRequestHandler {
  return (err, _req, res, _next) => {
    const errors: object[] = [];

    if (err instanceof ValidationError) {
      errors.push({
        message: err.message,
        field: err.field,
      });
    } else if (err instanceof ServerError) {
      const message = err.message ? err.message : DEFAULT_ERROR_MESSAGE;

      errors.push({ message });
    } else {
      errors.push({
        message: DEFAULT_ERROR_MESSAGE,
      });
    }

    const statusCode = err?.statusCode || 500;
    return res.status(statusCode).json({ errors });
  };
}
