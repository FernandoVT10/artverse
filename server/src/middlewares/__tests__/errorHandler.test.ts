import LoggerHandler from "@utils/LoggerHandler";
import errorHandler from "../errorHandler";
import mockExpress from "@test-utils/mockExpress";

import { ServerError, ValidationError } from "@utils/errors";

jest.mock("@utils/LoggerHandler");

describe("middlewares/checkValidation", () => {
  const callErrorHandler = (error: any = null) => {
    const expressMock = mockExpress();
    const { req, res, next } = expressMock;

    const handler = errorHandler();
    handler(error, req, res, next);

    return expressMock;
  };

  it("should call 'res.status' with code 500 by default", () => {
    const { res } = callErrorHandler();
    expect(res.status).toHaveBeenCalledWith(500);
  });

  it("should call 'res.json' with a default error message", () => {
    const { res } = callErrorHandler();
    const { errors } = res.json.mock.calls[0][0];
    expect(errors).toMatchSnapshot();
  });

  it("should call 'logError' with the error when the error is neither ServerError nor ValidationError", () => {
    const error = new Error();
    callErrorHandler(error);

    const logErrorMocked = jest.mocked(LoggerHandler.logError);
    expect(logErrorMocked).toHaveBeenCalledWith(error);
  });

  describe("ServerError instance", () => {
    it("should call 'res.status' with code 500", () => {
      const error = new ServerError();
      const { res } = callErrorHandler(error);
      expect(res.status).toHaveBeenCalledWith(500);
    });

    it("should call 'res.json' with a default error message when the ServerError message isn't given", () => {
      const error = new ServerError();

      const { res } = callErrorHandler(error);

      const { errors } = res.json.mock.calls[0][0];
      expect(errors).toMatchSnapshot();
    });

    it("should call 'res.json' with the error message of ServerError when is given", () => {
      const error = new ServerError("test");

      const { res } = callErrorHandler(error);

      const { errors } = res.json.mock.calls[0][0];
      expect(errors).toContainEqual({
        message: error.message,
      });
    });
  });

  describe("ValidationError instance", () => {
    const error = new ValidationError("foo", "test");

    it("should call 'res.status' with code 400", () => {
      const { res } = callErrorHandler(error);
      expect(res.status).toHaveBeenCalledWith(400);
    });

    it("should call 'res.json' with the error and field values of ValidationError", () => {
      const { res } = callErrorHandler(error);

      const { errors } = res.json.mock.calls[0][0];
      expect(errors).toContainEqual({
        message: error.message,
        field: error.field,
      });
    });
  });
});
