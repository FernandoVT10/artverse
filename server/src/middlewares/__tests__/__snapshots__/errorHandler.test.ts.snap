// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middlewares/checkValidation ServerError instance should call 'res.json' with a default error message when the ServerError message isn't given 1`] = `
Array [
  Object {
    "message": "There was an error trying to process your request. Please try again later.",
  },
]
`;

exports[`middlewares/checkValidation should call 'res.json' with a default error message 1`] = `
Array [
  Object {
    "message": "There was an error trying to process your request. Please try again later.",
  },
]
`;
